### PR TITLE
buildNpmPackage: make nodejs overridable

### PR DESCRIPTION
**Pull request by happysalada**
## Description of changes

Current nodejs stable in nixpkgs is 18.x , the upgrade to 20.x is approaching (most likely after the 23.11 release) but upstream LTS is 20.x.
I thought It could help maintainers make some upgrade to the packages they maintain by making the nodejs version overridable.

let me know if there are considerations I've missed of course.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@lilyinstarlight @winterqt @dotlambda 
This triggers 0 rebuilds and most likely only adds functionality.
let me know if I've missed anything, this should be a quick review.
(no worries if you're busy of course).

**Review by lilyinstarlight — COMMENTED** [review]
I'm in favor of this change, but could you add the argument to the buildNpmPackage docs?

Thank you!

It also might be worth doing a treewide change to update existing usages of `(buildNpmPackage.override { nodejs = ...; }) { ... }` too, but I won't make you do that in this PR :)

**Comment by happysalada** [comment]
@lilyinstarlight thank you for the comment!
I've made the change, let me know if the wording works.

I've checked and there are like 4 overrides of nodejs in nixpkgs, I'm happy to do this in a follow up PR if this helps.

My next PR if that works for everyone will be to add npmDeps as an argument as well. When building several packages off a workspace, it would speed up builds and reduce space usage to not have to build multiple dependencies version.

let me know if anything of course.

**Comment by lilyinstarlight** [comment]
> I've checked and there are like 4 overrides of nodejs in nixpkgs, I'm happy to do this in a follow up PR if this helps.

Up to you, but it would be appreciated!

> My next PR if that works for everyone will be to add npmDeps as an argument as well. When building several packages off a workspace, it would speed up builds and reduce space usage to not have to build multiple dependencies version.

You know, I actually thought that already worked, but unlike some of the args, it doesn't do a `args.npmDeps or ...` so it unconditionally overrides it in `buildNpmPackage`. An update to allow that to work would be welcome!

File: doc/languages-frameworks/javascript.section.md [hunk 1]
```diff
@@ -209,6 +209,7 @@ In the default `installPhase` set by `buildNpmPackage`, it uses `npm pack --json
 * `npmPackFlags`: Flags to pass to `npm pack`.
 * `npmPruneFlags`: Flags to pass to `npm prune`. Defaults to the value of `npmInstallFlags`.
 * `makeWrapperArgs`: Flags to pass to `makeWrapper`, added to executable calling the generated `.js` with `node` as an interpreter. These scripts are defined in `package.json`.
+* `nodejs`: The nodejs package to build the npm package. This in turns uses the corresponding npm package shipped with the nodejs chosen.
```
Comment (lilyinstarlight) [inline doc/languages-frameworks/javascript.section.md:212]:
```suggestion
* `nodejs`: The `nodejs` package to build against, using the corresponding `npm` shipped with that version of `node`. Defaults to `pkgs.nodejs`.
```

Maybe that might be slightly clearer?

Idk I'm open to alternative wordings too, it's just using "npm package" in two separate ways in this phrasing in confusing and it doesn't clearly state that it has a default value

Thanks again!

Comment (happysalada) [reply doc/languages-frameworks/javascript.section.md]: happy to go with that, I'll commit this then will just squash the commits into one upon merging if that works for you.

**Review by lilyinstarlight — APPROVED** [review]
Looks good now! Feel free to squash merge yourself

**Pull request merged**